File: apps/android/app/src/main/java/ai/drsai/remote/remote/data/WebSearchContract.kt
package ai.drsai.remote.remote.data

const val WEB_SEARCH_CONTRACT_SCHEMA = "opendrsai.web-search/1"
const val MANAGED_WEB_SEARCH_MODEL = "hepai/tavily-web-search-v1"

enum class WebSearchRecoveryAction {
    LOGIN, ACCOUNT_OR_BYOK, RETRY, RETRY_OR_BYOK, CONTACT_ADMIN, CHANGE_URL, EDIT_REQUEST,
}

data class WebSearchRecovery(val retryable: Boolean, val action: WebSearchRecoveryAction)

val WEB_SEARCH_FUNCTIONS = listOf("search", "extract")
val WEB_SEARCH_PROVIDER_MODES = listOf("auto", "managed", "byok", "none")
val WEB_SEARCH_RECOVERY = linkedMapOf(
    "login_required" to WebSearchRecovery(false, WebSearchRecoveryAction.LOGIN),
    "permission_denied" to WebSearchRecovery(false, WebSearchRecoveryAction.ACCOUNT_OR_BYOK),
    "rate_limited" to WebSearchRecovery(true, WebSearchRecoveryAction.RETRY),
    "quota_exhausted" to WebSearchRecovery(false, WebSearchRecoveryAction.ACCOUNT_OR_BYOK),
    "worker_unavailable" to WebSearchRecovery(true, WebSearchRecoveryAction.RETRY_OR_BYOK),
    "provider_authentication_failed" to WebSearchRecovery(false, WebSearchRecoveryAction.CONTACT_ADMIN),
    "provider_rate_limited" to WebSearchRecovery(true, WebSearchRecoveryAction.RETRY),
    "provider_quota_exhausted" to WebSearchRecovery(false, WebSearchRecoveryAction.CONTACT_ADMIN),
    "provider_timeout" to WebSearchRecovery(true, WebSearchRecoveryAction.RETRY),
    "provider_unavailable" to WebSearchRecovery(true, WebSearchRecoveryAction.RETRY),
    "provider_invalid_response" to WebSearchRecovery(true, WebSearchRecoveryAction.RETRY),
    "unsafe_web_url" to WebSearchRecovery(false, WebSearchRecoveryAction.CHANGE_URL),
    "invalid_request" to WebSearchRecovery(false, WebSearchRecoveryAction.EDIT_REQUEST),
)

fun webSearchRecovery(code: String?): WebSearchRecovery? = WEB_SEARCH_RECOVERY[code]

File: apps/ui-tui/src/webSearchContract.ts
/** Frozen public Web Search contract shared with Desktop and Android. */
export const WEB_SEARCH_CONTRACT_SCHEMA = 'opendrsai.web-search/1' as const
export const MANAGED_WEB_SEARCH_MODEL = 'hepai/tavily-web-search-v1' as const
export const WEB_SEARCH_FUNCTIONS = ['search', 'extract'] as const
export const WEB_SEARCH_PROVIDER_MODES = ['auto', 'managed', 'byok', 'none'] as const
export const WEB_SEARCH_ERROR_CODES = [
  'login_required', 'permission_denied', 'rate_limited', 'quota_exhausted',
  'worker_unavailable', 'provider_authentication_failed', 'provider_rate_limited',
  'provider_quota_exhausted', 'provider_timeout', 'provider_unavailable',
  'provider_invalid_response', 'unsafe_web_url', 'invalid_request',
] as const

export type WebSearchErrorCode = typeof WEB_SEARCH_ERROR_CODES[number]
export type WebSearchRecoveryAction = 'login' | 'account_or_byok' | 'retry' | 'retry_or_byok' | 'contact_admin' | 'change_url' | 'edit_request'

const RECOVERY: Readonly<Record<WebSearchErrorCode, { retryable: boolean; action: WebSearchRecoveryAction }>> = {
  login_required: { retryable: false, action: 'login' },
  permission_denied: { retryable: false, action: 'account_or_byok' },
  rate_limited: { retryable: true, action: 'retry' },
  quota_exhausted: { retryable: false, action: 'account_or_byok' },
  worker_unavailable: { retryable: true, action: 'retry_or_byok' },
  provider_authentication_failed: { retryable: false, action: 'contact_admin' },
  provider_rate_limited: { retryable: true, action: 'retry' },
  provider_quota_exhausted: { retryable: false, action: 'contact_admin' },
  provider_timeout: { retryable: true, action: 'retry' },
  provider_unavailable: { retryable: true, action: 'retry' },
  provider_invalid_response: { retryable: true, action: 'retry' },
  unsafe_web_url: { retryable: false, action: 'change_url' },
  invalid_request: { retryable: false, action: 'edit_request' },
}

export function webSearchRecovery(code: string): { retryable: boolean; action: WebSearchRecoveryAction } | null {
  return Object.prototype.hasOwnProperty.call(RECOVERY, code) ? RECOVERY[code as WebSearchErrorCode] : null
}
